### PR TITLE
use single command to uninstall mesh and cluster wide resources

### LINKED ISSUE #4429: Use a single command to for uninstall of mesh and cluster wide resources
**Please describe the Improvement and/or Feature Request**

There are 2 commands pertaining to uninstall:
1. `osm uninstall mesh` that uninstalls the control plane
2. `osm uninstall cluster-wide-resources` that uninstalls the cluster wide resources such as CRDs created after install

Having uninstall split across 2 commands seems unnecessary to me. This issue proposes to update the `osm uninstall` command to do the following:
1. `osm uninstall --all` to delete the mesh and cluster wide resources (equivalent to `osm uninstall mesh && osm uninstall cluster-wide-resources`
2. `osm uninstall` simply removes the mesh (equivalent to `osm uninstall mesh`)

**Scope (please mark with X where applicable)**
- CLI Tool                   [X]

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:

This commit combines the osm uninstall mesh and osm uninstall
cluster-wide-resources to a single command. The user can specify
deletion of cluster wide resources by using the flag
--delete-cluster-wide-resources.

If there is no mesh in the cluster, the user can still use the same
command to delete the cluster wide resources by providing the flag.

 This command has been made idempotent as
well.

fixes #4429

Signed-off-by: Sneha Chhabria <snchh@microsoft.com>

**Testing done**: Manually tested a bunch of scenarios and updated unit tests

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [X] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [X] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? `no`
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change? `no`

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)? `no will handle this next`

### CODE REVIEW

File: cmd/cli/uninstall_mesh.go [hunk 1]
```diff
@@ -18,40 +22,37 @@ import (
...
-	that CRD to also be deleted.
+(which OSM adds to support multiple CR versions)
 
-CRDs, mutating and validating webhooks, secrets, and the osm control
-plane namespace are left over when this command is run.
-
-Use 'osm uninstall cluster-wide-resources' to fully uninstall
-leftover osm resources from the cluster.
+The command will not delete:
+(1) the namespace the mesh was installed in unless specified via via the
```
Comment (shashankram) [inline cmd/cli/uninstall_mesh.go:33]:
```suggestion
(1) the namespace the mesh was installed in unless specified via the 
```

File: cmd/cli/uninstall_mesh.go [hunk 2]
```diff
@@ -80,70 +81,300 @@ func newUninstallMeshCmd(config *action.Configuration, in io.Reader, out io.Writ
...
 		},
 	}
 
 	f := cmd.Flags()
 	f.StringVar(&uninstall.meshName, "mesh-name", defaultMeshName, "Name of the service mesh")
 	f.BoolVarP(&uninstall.force, "force", "f", false, "Attempt to uninstall the osm control plane instance without prompting for confirmation.  If the control plane with specified mesh name does not exist, do not display a diagnostic message or modify the exit status to reflect an error.")
+	f.BoolVarP(&uninstall.deleteClusterWideResources, "delete-cluster-wide-resources", "a", false, "Cluster wide resources (such as osm CRDs, mutating webhook configurations, validating webhook configurations and osm secrets) are fully deleted from the cluster after control plane components are deleted.")
 	f.BoolVar(&uninstall.deleteNamespace, "delete-namespace", false, "Attempt to delete the namespace after control plane components are deleted")
 	f.Uint16VarP(&uninstall.localPort, "local-port", "p", constants.OSMHTTPServerPort, "Local port to use for port forwarding")
+	f.StringVar(&uninstall.caBundleSecretName, "ca-bundle-secret-name", constants.DefaultCABundleSecretName, "Name of the secret for the OSM CA bundle")
```
Comment (shashankram) [inline cmd/cli/uninstall_mesh.go:100]: The CA bundle secret will be removed by the cleanup hook. So wondering why is this required.

Comment (snehachhabria) [reply cmd/cli/uninstall_mesh.go:100]: This is for scenarios when the cleanup hook isn't used

**Review by johnsonshi — APPROVED** [review]

File: cmd/cli/uninstall_mesh.go [hunk 3]
```diff
@@ -80,70 +81,303 @@ func newUninstallMeshCmd(config *action.Configuration, in io.Reader, out io.Writ
...
+				}
+				return errors.Errorf("Error occurred while deleting OSM namespace [%s] - %v", d.meshNamespace, err)
+			}
+			fmt.Fprintf(d.out, "OSM namespace [%s] deleted successfully\n", d.meshNamespace)
+		}
+	} else {
+		if d.deleteNamespace {
+			fmt.Fprintf(d.out, "OSM namespace [%s] CANNOT be deleted in a managed environment\n", d.meshNamespace)
+		}
+	}
```
Comment (jaellio) [inline cmd/cli/uninstall_mesh.go:196]: Just a suggestion :)

```suggestion
	if d.deleteNamespace {
		 if !settings.IsManaged() {
			if err = d.clientSet.CoreV1().Namespaces().Delete(ctx, d.meshNamespace, v1.DeleteOptions{}); err != nil {
				if k8sApiErrors.IsNotFound(err) {
					fmt.Fprintf(d.out, "OSM namespace [%s] not found\n", d.meshNamespace)
					return nil
				}
				return errors.Errorf("Error occurred while deleting OSM namespace [%s] - %v", d.meshNamespace, err)
			}
			fmt.Fprintf(d.out, "OSM namespace [%s] deleted successfully\n", d.meshNamespace)
		} else {
			fmt.Fprintf(d.out, "OSM namespace [%s] CANNOT be deleted in a managed environment\n", d.meshNamespace)
		}
	}
```

File: cmd/cli/uninstall_mesh.go [hunk 4]
```diff
@@ -80,70 +81,303 @@ func newUninstallMeshCmd(config *action.Configuration, in io.Reader, out io.Writ
...
+
+			if !d.deleteClusterWideResources && !d.deleteNamespace {
+				return err
+			}
+		}
+
+		if err == nil {
+			fmt.Fprintf(d.out, "OSM [mesh name: %s] in namespace [%s] uninstalled\n", d.meshName, d.meshNamespace)
+		}
+	}
```
Comment (jaellio) [inline cmd/cli/uninstall_mesh.go:151]: If the uninstall cmd is being applied to a managed environment, can you add a message to the output indicating that the mesh cannot be uninstalled?
```suggestion
	} else {
	    fmt.Fprintf(d.out, "OSM [mesh name: %s] in namespace [%s] CANNOT be uninstall in a managed environment", d.meshName, d.meshNamespace)
	}
```

File: cmd/cli/uninstall_mesh.go [hunk 5]
```diff
@@ -80,70 +81,303 @@ func newUninstallMeshCmd(config *action.Configuration, in io.Reader, out io.Writ
...
 		}
-		fmt.Fprintf(d.out, "OSM namespace [%s] deleted successfully\n", ns)
 	}
 
-	return err
+	return nil
+}
+
+// validateCLIParams contains all checks necessary to ensure that various permutations of the CLI flags are consistent
+func (d *uninstallMeshCmd) validateCLIParams() error {
```
Comment (shashankram) [inline cmd/cli/uninstall_mesh.go:202]: Since there are defaults for these flags, why is this validation necessary?

**Review by shashankram — APPROVED** [review]

File: cmd/cli/uninstall_mesh.go (hunk 5 above)
Comment (snehachhabria) [reply cmd/cli/uninstall_mesh.go]: good catch, it's not necessary removed it

**Review by jaellio — APPROVED** [review]
